<!DOCTYPE html>
<html lang="en">
<head>
	<title>Snakemake—simple data processing for researchers</title>
	<meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, viewport-fit=cover">
	<link rel="stylesheet" href="node_modules/@highlightjs/cdn-assets/styles/github.min.css" type="text/css"/>
	<link rel="stylesheet" href="styles/styles.css">
    <style>
        /* Customize the theme */

		.shower {
			--slide-ratio: calc(16 / 9);
			--color-key: #9b582f;
			--color-white: #eaeaea;
			--slide-notes-max-width: 80ch;
		}


		/* Explicit notes control labels */

		.control {
			--radius: 0.75em;
			--line-height: 1.5em;
		}

		.control input[type='checkbox'] + label::after {
			text-align: center;
			content: "N";
			color: var(--color-off);
		}

		.control input[type='checkbox']:checked + label::after {
			content: "Y";
			color: var(--color-on);
		}
    </style>
</head>
<body class="shower list show-notes">

	<header class="caption">
		<h1>Snakemake—simple data processing for researchers</h1>
		<p><a href="https://blog.liang2.tw/">Liang-Bo Wang</a>, 2022-09-17.</p>
	</header>

	<menu class="control">
		<li class="show-notes">
			<input type="checkbox" id="show-notes-toggle" checked>
			<label for="show-notes-toggle">Slide notes</label>
		</li>
	</menu>

	<!-- Cover Slide -->
	<section id="cover" class="slide clear black">
		<h2 id="talk-header">Snakemake—simple data processing for researchers</h2>
		<p id="talk-info">
			Liang-Bo Wang<br>
			Taipei Bioinformatics Omnibus<br>
			<span style="color: var(--color-black)">2022-09-17</span>
		</p>
		<p id="usage-instr">
			<kbd>Esc</kbd> to exit<br>
			<kbd>←</kbd> <kbd>→</kbd> to navigate<br>
			<a rel="license" href="http://creativecommons.org/licenses/by/4.0/">
				<img alt="Creative Commons Attribution 4.0 International License"
					src="pictures/cc-by-4.0.svg"/>
			</a>
		</p>
		<figure>
			<img src="pictures/donggeun-lee-BpWvu_KLJxo-unsplash.jpg" class="cover w" alt="">
			<figcaption class="copyright bottom white">
				Photo by <a href="https://unsplash.com/@dizy64">DongGeun Lee</a>
				on <a href="https://unsplash.com/photos/BpWvu_KLJxo">Unsplash</a>
			</figcaption>
		</figure>
		<style>
			#talk-header, #talk-info {
				left: var(--slide-left-side);
				right: var(--slide-right-side);
				text-align: center;
				position: absolute;
				color: var(--color-white);
				margin: 0;
			}
			#talk-header {
				font-size: 3rem;
				font-weight: 600;
				line-height: 1.25;
				top: 8rem;
				/* CSS Gradient https://cssgradient.io/ */
				background-image: linear-gradient(
					to right,
					#9b582f5a 0%,
					#4f453c8e 100%
				);
				left: 0;
				right: 0;
				padding: 1.25em calc(var(--slide-left-side) / 2);
			}
			#talk-info {
				font-size: 1.5rem;
				line-height: 1.5;
				bottom: 4rem;
			}
			#cover .copyright {
				left: calc(var(--slide-left-side) / 2);
			}
			#usage-instr {
				color: var(--color-white);
				font-size: 1rem;
				line-height: 1.75;
				right: calc(var(--slide-right-side) / 2);
				bottom: 20px;
				position: absolute;
				text-align: right;
				margin: 0;
			}
			#usage-instr a[rel="license"] {
				background-image: none;
			}
			#usage-instr a[rel="license"] img {
				height: 1em;
				color: var(--color-white);
			}
			#usage-instr kbd {
				color: var(--color-black);
				background-color: var(--color-light);
			}
		</style>
	</section>

	<section id="target-audience" class="slide">
		<h2>This talk is for researchers who need to automate things in a handy way</h2>
		<div class="two columns l67">
			<ul>
				<li>Some advices may not apply to a bioinformatics core lab, who develop pipelines as their main job</li>
				<li class="next">This is <em>not</em> a complete Snakemake tutorial;<br>
					please learn it properly from the <a href="https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html">official tutorial</a></li>
				<li class="next">Share my experience of Snakemake and how to best utilize it</li>
			</ul>
			<figure>
				<img
					class="fill w"
					src="pictures/beesy_researcher_DALL·E.png"
					alt="DALL·E generated art using the prompt 'a hard-working bee working on multiple tasks, digital illustration'">
				<figcaption class="citation right">A bee-sy dry-lab researcher</figcaption>
			</figure>
		</div>
	</section>

	<section id="scenarios" class="slide tight">
		<h2>My top needs for a pipeline/workflow</h2>
		<div class="two columns">
			<div>
				<p>To process multiple samples on different computing platforms</p>
				<figure>
					<img class="fill w"
						src="pictures/scenario-data-processing_DALL·E.png"
						alt="DALL·E generated art using the prompt 'four repeats in a two by two grid of a robot assembling car parts on a conveyor belt, digital art'">
				</figure>
			</div>
			<div>
				<p>To reproduce my analysis steps and results</p>
				<figure>
					<img class="fill w"
						src="pictures/secnario-reproducible-analysis_DALL·E.png"
						alt="DALL·E generated art using the prompt 'a robot arm handling multiple sheets of papers with scientific charts and illustrations on a wood table, digital art'">
				</figure>
			</div>
		</div>
		<style>
			#scenarios figure {
				width: 20rem;
				margin-top: 0.5em;
			}
		</style>
	</section>

	<section id="scenario-data-processing" class="slide">
		<h2>Scenario 1: A pipeline to process multiple samples on different computing platforms</h2>
		<div class="two columns l67">
			<ul>
				<li>Run at different environments: standalone server, HPC cluster, and cloud</li>
				<li>HPC cluster has its own job queue and computing architecture (<a href="https://en.wikipedia.org/wiki/IBM_Spectrum_LSF">LSF</a> at my school)</li>
				<li>We use public cloud services (Google Cloud and AWS) and we manage the usage ourselves</li>
			</ul>
			<figure>
				<img
					class="fill w"
					src="pictures/scenario-data-processing_DALL·E.png"
					alt="Illustration of the scenario of a reproducible analysis workflow. DALL·E generated art.">
			</figure>
		</div>
		<style>
			#scenario-data-processing figure {
				margin-top: 0.5em;
			}
		</style>
	</section>

	<section class="slide-notes">
		<h2>Scenario 1: Scalable data processing pipeline</h2>
	</section>

	<section id="scenario-reproducible-analysis" class="slide">
		<h2>Scenario 2: A workflow to reproduce my analysis steps and results</h2>
		<div class="two columns l67">
			<ul>
				<li>Steps to build the dataset for downstream analysis</li>
				<li>Figure generations in the manuscript</li>
				<li>Update the results when the new data come in</li>
				<li>Preferrably, others can easily set up the workflow and reproduce in an unknown environment</li>
			</ul>
			<figure>
				<img
					class="fill w"
					src="pictures/secnario-reproducible-analysis_DALL·E.png"
					alt="Illustration of the scenario of a reproducible analysis workflow. DALL·E generated art.">
			</figure>
		</div>
		<style>
			#scenario-reproducible-analysis figure {
				margin-top: 0.5em;
			}
		</style>
	</section>

	<section class="slide-notes">
		<h2>Scenario 2: Reproducible data analysis workflow</h2>
	</section>

	<section id="scenario-comparison" class="slide tight">
		<table>
			<thead>
				<tr>
					<th></th>
					<th scope="col">
						<img
						src="pictures/scenario-data-processing_DALL·E.png"
						alt="Illustration of the scenario of a reproducible analysis workflow. DALL·E generated art."><br>
						Data processing
					</th>
					<th scope="col">
						<img
						src="pictures/secnario-reproducible-analysis_DALL·E.png"
						alt="Illustration of the scenario of a reproducible analysis workflow. DALL·E generated art."><br>
						Reproducible analysis
					</th>
				</tr>
			</thead>
			<tbody>
				<tr>
					<th scope="row">Inputs</th>
					<td>Lots of samples</td>
					<td>Mostly the same inputs</td>
				</tr>
				<tr>
					<th scope="row">Run type</th>
					<td>Once per sample</td>
					<td>Multiple times</td>
				</tr>
				<tr>
					<th scope="row">Workflow stability</th>
					<td>Stable</td>
					<td>Subject to change</td>
				</tr>
				<tr>
					<th scope="row">Computing environment</th>
					<td>Known and multiple<br>
						local, HPC, and cloud</td>
					<td>Known and limited*<br>
						mostly local</td>
				</tr>
			</tbody>
		</table>
		<style>
			#scenario-comparison img {
				height: 7em;
			}
		</style>
	</section>

	<section id="about-me" class="slide">
		<h2>I'm Liang-Bo Wang</h2>
		<div class="two columns r33">
			<ul>
				<li>PhD in Computational and Systems Biology<br>
					from Washington University in St. Louis</li>
				<li>My thesis focuses on cancer genomics</li>
				<li>Worked in related consortia including
					<a href="https://proteomics.cancer.gov/programs/cptac">CPTAC</a>,
					<a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">TCGA</a>,
					and <a href="https://humantumoratlas.org/">HTAN</a></li>
				<li>Common tasks as a researcher: prototyping new tools/pipelines, data wrangling, and analyses</li>
			</ul>
		</div>
		<img
			class="fill h place r"
			src="pictures/my_protrait.jpg"
			alt="My protrait">
		<style>
			#about-me abbr {
				font-variant-caps: normal;
			}
		</style>
	</section>

	<section id="what-is-snakemake" class="slide">
		<img src="pictures/snakemake_logo.svg" alt="Logo of Snakemake">
		<h2>What is Snakemake?</h2>
		<p><a href="https://snakemake.github.io/">Snakemake</a> is a Python-based Makefile-like workflow management tool.<br>
			A workflow contains a set of <strong>rules</strong>:</p>
		<ul>
			<li>A rule describes how to create output files/patterns from input files/pattern</li>
			<li>The workflow file runs like a Python script</li>
			<li>Can import external Python libraries</li>
			<li>Can create one's own Python functions</li>
		</ul>
		<style>
			#what-is-snakemake img {
				float: right;
				width: 4.5em;
			}
		</style>
	</section>

	<section id="how-snakemake-works" class="slide tight">
		<h2>How Snakemake works in general</h2>
		<img src="pictures/example-dag.svg" alt="">
		<ol>
			<li>Snakemake finds the rule that generates the final outputs. <br>
				<mark>Trace the workflow backwards</mark></li>
			<li>If all the inputs exist, it's ready to run the commands of this rule<br>
				(create a new job)</li>
			<li>If some inputs do not exist, find other rules to generate the required input files (create a dependent job)</li>
			<li>Repeat the process until all outputs can be generated from existing files (complete a job dependency graph)</li>
			<li>Snakemake will start running the corresponding commands/jobs</li>
		</ol>
		<style>
			#how-snakemake-works img {
				float: right;
				margin-left: 0.5em;
			}
		</style>
	</section>

	<section id="snakemake-in-action" class="slide">
		<h2>A RNA-seq pipeline in action</h2>
		<p>Generate readcounts per gene from stranded RNA-seq aligned BAMs</p>
		<ol>
			<li>Find the local path of the sample BAMs</li>
			<li>Generate read counts (featureCounts)</li>
			<li>Clean up and compress the outputs</li>
			<li>Calculate FPKM, FPKM-UQ, and TPM values</li>
		</ol>
	</section>

	<section id="example-rule" class="slide tight">
		<pre class="language-snakemake">
<code>rule readcount:
    output: count_tsv='stranded_readcount/{sample}.tsv'
    input: bam='inputs/{sample}.bam',
		   bai='inputs/{sample}.bam.bai'
    log: 'logs/stranded_readcount/{sample}.log'
    params: gtf=GENE_GTF_PTH
    threads: 16
    shell:
        'featureCounts -T {threads} -a {params.gtf} ...'
        '-o {output.count_tsv} {input.bam} 2> {log}'</code>
		</pre>
	</section>

	<section id="example-notes" class="slide-notes">
		<h2>Snakemake syntax explained</h2>
		<dl>
			<dt><code>{sample}</code></dt>
			<dd>
				Known as <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#wildcards">wildcards</a> that matches the possible existing files.
			</dd>
			<dt><code>input:</code>, <code>output:</code></dt>
			<dd>File patterns for inputs and outputs. Use key-value pairs <code><strong>key</strong>='...'</code> when there are multiple patterns. They are just Python strings and thus can be referred elsewhere by <code>rules.<strong>myrule</strong>.output.<strong>key</strong></code>.</dd>
			<dt></dt>
			<dd></dd>
		</dl>
		<p>Due to the way snakemake executes, it's easier to resolve the output first by realizing all the wildcard variables. Then expands the rest of the rule (inputs, parameters, and etc.)</p>

		<p>For example, if there is a output <code>stranded_readcount/<mark>my_sample_a</mark>.tsv</code>, <code>{sample}</code> expands to <code>my_sample_a</code> in all occurrences. So snakemake will look for the corresponding inputs and generate the corresponding shell commands:</p>
		<pre class="language-shell">
<code>featureCounts -T 16 -a /path/to/genes.gtf \
	-o stranded_readcount/my_sample_a.tsv \
	inputs/my_sample_a.bam \
	2> logs/stranded_readcount/my_sample_a.log</code></pre>
		<p>To make the pipeline more extensible, use <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html">a config file</a> and a run-all rule to process all samples of a batch.</p>

		<style>
			#example-notes {
				grid-row-end: span 2;
			}
		</style>
	</section>

	<section id="example-execution" class="slide tight">
		<pre class="language-bash">
<code>$ snakemake -j 32 stranded_readcount/my_sample_{a,b,c,d}.tsv
# featureCounts ... inputs/my_sample_a.bam
# featureCounts ... inputs/my_sample_b.bam
# ... (two jobs in parallel)</code>
		</pre>
		<p class="next">Alternatively, add a run-all rule to <code>snakemake all</code>:</p>
		<pre class="language-snakemake next">
<code>with open(config['sample_list']) as f:
	SAMPLES = f.read().splitlines()
rule all:
	input: expand(rules.readcount.output.count_tsv, \
				  sample=SAMPLES)</code>
		</pre>
	</section>

	<section id="example-rule-extended" class="slide tight">
		<p>Other powerful features to go beyond wildcard expansion:</p>
		<pre class="language-snakemake">
<code>rule readcount:
    input: unpack(find_bam_path)  # fills in bam and bai
    resources:  # dynamic resource allocation
		mem_mb=lambda wildcards, attempt: \
			16000 + 16000 * (attempt - 1)

def find_bam_path(wildcards):
	bam_pth = f"{config['bam_dir']}/{wildcards.sample}.bam"
	return {'bam': bam_pth, 'bai': bam_pth + '.bai'}</code>
		</pre>
	</section>

	<section id="things-i-like" class="slide tight">
		<h2>Things I like about Snakemake</h2>
		<ul>
			<li>Separate the definition and execution of tasks
				<ul>
					<li>Maximally utilize the available computing resources (CPUs and memory)</li>
					<li>Smart re-run on only the incomplete tasks (for iterative development)</li>
					<li>Integration with external job schedulers (e.g., HPC and Kubernetes)</li>
				</ul>
			</li>
			<li class="next">Powerful features that are easy to use
				<ul>
					<li>Flexible rules by custom Python functions</li>
					<li>Temporary and piped outputs</li>
					<li>Package management (conda),
						containerization (docker),
						and tool wrappers</li>
					<li>New features are <a href="https://snakemake.readthedocs.io/en/stable/project_info/history.html">constantly being introduced</a></li>
				</ul>
			</li>
		</ul>
	</section>

	<section id="things-i-dont-like" class="slide tight">
		<h2>Things I don't like about Snakemake</h2>
		<ul>
			<li>Running large number of jobs (> 100k) can be slow/fragile</li>
			<li>Flexible features can be a double-edged sword; not the most robust
				<ul>
					<li>Unique combinations of features can trigger weird bugs or crash the runtime</li>
					<li>Though bugs are usually fixed eventually, it may take up too much of the development time for a researcher</li>
				</ul>
			</li>
			<li>Large setup & runtime overhead to run on cloud
				<ul>
					<li>All files need to be tracked (e.g., inputs, outputs, and references)</li>
					<li>Overhead in downloading/uploading files and setting up the environment</li>
					<li>No one simple workflow for all setups; require optimization</li>
				</ul>
			</li>
		</ul>
	</section>

	<section id="best-practices-for-data-processing" class="slide tight">
		<img
			class="illustration place top right"
			src="pictures/scenario-data-processing_DALL·E.png"
			alt="Illustration of the scenario of a reproducible analysis workflow. DALL·E generated art.">
		<h2>My best practices for data processing</h2>
		<ul>
			<li>Run samples in batches; use config file</li>
			<li>Use file checksums to track upstream sources and outputs</li>
			<li>A big fat Docker/conda environment to contain all dependencies*</li>
			<li>Do I really need a "pipeline"? (run for 10+ times) <br>
				Good-old bash scripts plus <a href="https://www.gnu.org/software/parallel/sphinx.html">GNU Parallel</a> may be sufficient or more worth the time
			</li>
			<li>Running on cloud will be difficult than local/HPC unless there is a team maintaining the cloud infrastructure</li>
		</ul>
		<style>
			#best-practices-for-data-processing .illustration {
				width: 10rem;
			}
		</style>
	</section>

	<section id="best-practices-for-reproducible-analysis" class="slide tight">
		<img
			class="illustration place top right"
			src="pictures/secnario-reproducible-analysis_DALL·E.png"
			alt="Illustration of the scenario of a reproducible analysis workflow. DALL·E generated art.">
		<h2>My best practices for reproducible analysis</h2>
		<ul>
			<li>Start with making the independent reproducible scripts<br>
				(e.g., R Markdown/Jupyter notebooks)</li>
			<li>Construct the workflow in stages:
				<ol>
					<li>Make a data freeze with change log and checksums</li>
					<li>Run the analyses and number them (e.g. <code>09_xxx.Rmd</code>)</li>
					<li>Generate figures/tables/number for publication</li>
				</ol>
			</li>
			<li>Regularly test the workflow</li>
			<li>Work on the workflow later when the project is in shape (refactoring takes time)</li>
		</ul>
		<style>
			#best-practices-for-reproducible-analysis h2 {
				font-size: 42px;
			}
			#best-practices-for-reproducible-analysis .illustration {
				width: 10rem;
			}
		</style>
	</section>

	<section id="snakemake-vs-cwl-nextflow" class="slide tight">
		<h2>Snakemake vs <abbr title="Common Workflow Language">CWL</abbr>/<abbr title="Workflow Description Language">WDL</abbr> (and Nextflow)</h2>
		<p><a href="https://www.commonwl.org/">CWL</a>, <a href="https://openwdl.org/">WDL</a>, and <a href="https://www.nextflow.io/">Nextflow</a> ...</p>
		<ul>
			<li>Based on an explicit and forward task dependency (push-based <abbr title="Directed Acyclic Graph">DAG</abbr>)</li>
			<li>Generate <strong>unique IDs per workflow run</strong>; useful for large scale processing, monitoring, and integration with <abbr title="Laboratory Information Management System">LIMS</abbr></li>
			<li><strong>Cloud friendly</strong>; support diverse backends</li>
			<li>Robust, but less flexible and more tedious to write</li>
			<li>Ideal for a core lab and a workflow that will run for 100+ times</li>
			<li>Not very useful for reproducing analyses/figures</li>
		</ul>
		<style>
			#snakemake-vs-cwl-nextflow h2 {
				font-size: 42px;
			}
		</style>
	</section>

	<section id="when-to-use-snakemake" class="slide">
		<h2>When to use Snakemake or others?</h2>
		<ul>
			<li>If you already know Python, new to workflow, or don't develop pipelines for a living, pick Snakemake.</li>
			<li>For prototyping and managing your research results, also pick Snakemake</li>
			<li>For pipelines that are critical or run on cloud, pick others</li>
		</ul>
		<p>CWL/WDL is the most popular approach in my community (used by <a href="https://github.com/NCI-GDC/gdc-workflow-overview">NCI GDC</a>, <a href="https://terra.bio/">Terra</a>, and my institution), while Nextflow has a strong community promoting it and maintains a collection of <a href="https://nf-co.re/">ready-to-use workflows</a>.</p>
	</section>

	<section id="snakemake-vs-airflow" class="slide">
		<h2>Workflow engine choices</h2>
		<div class="columns two">
			<div>

			</div>
			<div>
				<figure>
					<img class="fill w"
						src="pictures/BenSiranosian-tweet.png"
						alt="Tweet screenshot by @BenSiranosian">
					<figcaption class="citation">
						From <a href="https://twitter.com/BenSiranosian/status/1558631901346164736">@BenSiranosian's tweet</a> and <a href="https://www.bsiranosian.com/bioinformatics/why-are-bioinformatics-workflows-different/">his blog post</a>.
					</figcaption>
				</figure>
			</div>
		</div>
		<style>
			#date-eng-vs-bioinfo figure {
				text-align: center;
			}
			#date-eng-vs-bioinfo figure img {
				width: 100%;
			}
		</style>
	</section>

	<section id="summary" class="slide">
		<h2>Summary</h2>
	</section>

	<div class="progress"></div>

	<!-- GitHub badge -->
	<footer class="badge">
		<a href="https://github.com/ccwang002/2022Talk-snakemake/">
			<svg width="81" height="79" viewBox="0 0 81 79" aria-hidden="true">
				<path d="M40.5 0a40.5 40.5 0 0 0-12.8 78.93c2 .37 2.76-.88 2.76-2s0-3.51-.05-6.89c-11.27 2.45-13.64-5.43-13.64-5.43-1.84-4.68-4.5-5.92-4.5-5.92-3.68-2.51.28-2.46.28-2.46 4.06.29 6.2 4.17 6.2 4.17 3.61 6.19 9.48 4.4 11.79 3.37a8.65 8.65 0 0 1 2.57-5.41c-9-1-18.45-4.5-18.45-20a15.65 15.65 0 0 1 4.17-10.87 14.57 14.57 0 0 1 .4-10.72s3.4-1.09 11.14 4.15a38.39 38.39 0 0 1 20.28 0c7.73-5.24 11.13-4.15 11.13-4.15a14.55 14.55 0 0 1 .4 10.72 15.63 15.63 0 0 1 4.16 10.87c0 15.56-9.47 19-18.49 20 1.45 1.25 2.75 3.72 2.75 7.5v11.11c0 1.08.73 2.34 2.78 1.95A40.51 40.51 0 0 0 40.5 0z"/>
			</svg>
			See related materials<br>
			on GitHub
		</a>
	</footer>

	<script src="node_modules/@shower/core/dist/shower.js"></script>
	<script defer src="node_modules/@highlightjs/cdn-assets/highlight.min.js"></script>
	<script defer src="js/highlightjs-snakemake.js"></script>
	<script>
		document.addEventListener('DOMContentLoaded', (event) => {
			// Display or hide the slide notes
			const body = document.querySelector('body');
			const notesToggle = document.querySelector('.control .show-notes input');
			notesToggle.checked ? body.classList.add('show-notes') : body.classList.remove('show-notes');
			notesToggle.addEventListener('click', (event) => {
				body.classList.toggle('show-notes');
			});

			// Load Highlight.js
			hljs.registerLanguage('snakemake', snakemake);
			hljs.highlightAll();
		});
	</script>
</body>
</html>
